src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java

```diff
@@ -253,6 +247,35 @@ private static ModuleFinder ofModuleInfos() {
         return new SystemModuleFinder(mrefs, nameToModule);
     }
 
+    /**
+     * Parses the {@code module-info.class} of all modules in the runtime image and
+     * returns a stream of {@link ModuleInfo.Attributes Attributes} for them. The
+     * returned attributes are in no specific order.
+     */
+    private static Stream<ModuleInfo.Attributes> allModuleAttributes() {
+        // System reader is a singleton and should not be closed by callers.
+        ImageReader reader = SystemImage.reader();
+        try {
+            return reader.findNode("/modules").getChildNames().map(mn -> readModuleAttributes(reader, mn));
+        } catch (IOException e) {
+            throw new Error("Error reading root /modules entry", e);
+        }
+    }
+
+    // Every module is required to have a valid module-info.class.
```

Comment:
The methods in this class use use /** ..*/ in the method descriptions so would prefer if the changes change that. Here the method should really say that it returns the module's module-info, returning a holder for its class file attributes.